pkg/database/prepared.go

```diff
@@ -0,0 +1,168 @@
+package database
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"sync"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const otelName = "prepared"
+
+var otelTracer trace.Tracer
+
+func init() {
+	otelTracer = otel.Tracer(otelName)
+}
+
+type preparedDb struct {
+	underlying Interface
+	mu         sync.RWMutex
+	store      map[string]*sql.Stmt
+}
+
+func NewPrepared(db Interface) Interface {
+	return &preparedDb{
+		underlying: db,
+		store:      make(map[string]*sql.Stmt),
+	}
+}
+
+func (db *preparedDb) ExecContext(ctx context.Context, query string, args ...any) (result sql.Result, err error) {
+	ctx, span := otelTracer.Start(ctx, "DB.ExecContext")
+	defer func() {
+		span.RecordError(err)
+		span.End()
+	}()
+
+	stmt, err := db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	return stmt.ExecContext(ctx, args...)
+}
+
+func (db *preparedDb) QueryContext(ctx context.Context, query string, args ...any) (rows *sql.Rows, err error) {
+	ctx, span := otelTracer.Start(ctx, "DB.QueryContext")
+	defer func() {
+		span.RecordError(err)
+		span.End()
+	}()
+
+	stmt, err := db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	return stmt.QueryContext(ctx, args...)
+}
+
+func (db *preparedDb) PrepareContext(ctx context.Context, query string) (stmt *sql.Stmt, err error) {
+	return db.underlying.PrepareContext(ctx, query)
+}
+
+func (db *preparedDb) prepare(ctx context.Context, query string) (stmt *sql.Stmt, err error) {
+	ctx, span := otelTracer.Start(ctx, fmt.Sprintf("%s.prepare", otelName))
+	defer func() {
+		span.RecordError(err)
+		span.End()
+	}()
+	span.SetAttributes(attribute.String("query", query))
```

Comment:
```suggestion
```
Let's remove the query, we don't want to expose any query data to the Otel endpoint.